experimental: Per service lb algorithm selection

DamianSawicki opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

---
This adds support for Per-service Load Balancing Algorithm Selection (from [CFP-34577](https://github.com/cilium/design-cfps/blob/main/cilium/CFP-34577-per-service-lb-algorithm-selection.md)) to StateDB-based load balancing after it was implemented in the currently-used load balancing in https://github.com/cilium/cilium/pull/35735/ (see also https://github.com/cilium/cilium/pull/36204).

kl52752 commented on pkg/loadbalancer/experimental/bpf_reconciler.go:
```diff
@@ -717,6 +719,19 @@ func (ops *BPFOps) updateFrontend(fe *Frontend) error {
 
+func (ops *BPFOps) shouldUseMaglev(fe *Frontend) bool {
+	if ops.cfg.LoadBalancerAlgorithmAnnotation {
+		if rawValue, ok := fe.service.Annotations[annotation.ServiceLoadBalancingAlgorithm]; ok {
```
there is a function checking the annotation in pkg/k8s/service.go:79, can you reuse it? Maybe it should be moved to some shared library?

kl52752 replied:
The value from the annotation should be set on Service fields (svcValue) so we can put it into bpf maps. I bealive it should be done in `BPFOps::updateFrontend` function

kl52752 replied:
New field `LoadBalancerAlgo loadbalancer.SVCLoadBalancingAlgo` should be added to https://github.com/cilium/cilium/blob/7530bcce954301a771ac603ed68b9accba3357a4/pkg/loadbalancer/experimental/service.go#L27 and handled in upsert and copy functions. 
Please see https://github.com/cilium/cilium/pull/35735/files#diff-8c03c282d14585836151157c0a5655f8192d35d39b65215c37794a8b63de7690

and comment from @joamaki https://github.com/cilium/cilium/pull/35735#discussion_r1851981426

DamianSawicki replied:
> there is a function checking the annotation in pkg/k8s/service.go:79, can you reuse it? Maybe it should be moved to some shared library?

I can't reuse it directly because it takes `*slim_corev1.Service` as argument (and we have `experimental.Service` here), but I reworked it a little bit to make it reusable, PTAL.

> The value from the annotation should be set on Service fields (svcValue) so we can put it into bpf maps. I bealive it should be done in `BPFOps::updateFrontend` function

Good catch, added!

> New field `LoadBalancerAlgo loadbalancer.SVCLoadBalancingAlgo` should be added to
> 
> https://github.com/cilium/cilium/blob/7530bcce954301a771ac603ed68b9accba3357a4/pkg/loadbalancer/experimental/service.go#L27
> 
> and handled in upsert and copy functions.

I was thinking about it, but I came to the conclusion that this would basically duplicate the information in the annotation, so it find it redundant in the StateDB loadbalancing context. The situation in `pkg/datapath/types/lbmap.go` is different because `UpsertServiceParams` does not contain the annotation field that is available in `experimental.Service`. This seems to agree with @joamaki's opinion

> Service object already carries the annotations, so nothing to do in service.go or reflector.go.

from the comment you linked.

> and comment from @joamaki [#35735 (comment)](https://github.com/cilium/cilium/pull/35735#discussion_r1851981426)

Thanks a lot for this reference! It seems the only missing point now is the following: 
> Though we probably should show the special annotations in the TableRow method so it's visible when looking at the tables.
> add the necessary code to DumpLBMaps to show the algorithm

kl52752 replied:
thanks looks good now!

kl52752 commented on pkg/loadbalancer/experimental/config.go:
```diff
@@ -40,28 +40,33 @@ type TestConfig struct {
 func (def TestConfig) Flags(flags *pflag.FlagSet) {
 	flags.Float32("lb-test-fault-probability", def.TestFaultProbability, "Probability for fault injection in LBMaps")
 	flags.String("node-port-algorithm", option.NodePortAlgRandom, "NodePort algorithm")
+	flags.Bool("bpf-lb-algorithm-annotation", false, "Enable service-level annotation for configuring BPF load balancing algorithm")
```
you set this flag always false for tests?

DamianSawicki replied:
It's just the default value. The actual values are passed [via](https://github.com/cilium/cilium/blob/2572a268039e5ee5fefd8dbc1de664091e426bee/pkg/loadbalancer/experimental/testdata/nodeport-explicit-maglev.txtar#L1) [flags](https://github.com/cilium/cilium/blob/2572a268039e5ee5fefd8dbc1de664091e426bee/pkg/loadbalancer/experimental/testdata/nodeport-explicit-random.txtar#L1) to [TestConfig](https://github.com/cilium/cilium/blob/2572a268039e5ee5fefd8dbc1de664091e426bee/pkg/loadbalancer/experimental/config.go#L44) and further to [DaemonConfig](https://github.com/cilium/cilium/blob/2572a268039e5ee5fefd8dbc1de664091e426bee/pkg/loadbalancer/experimental/script_test.go#L72) and [ExternalConfig](https://github.com/cilium/cilium/blob/2572a268039e5ee5fefd8dbc1de664091e426bee/pkg/loadbalancer/experimental/config.go#L70).

kl52752 commented on pkg/maps/lbmap/ipv4.go:
```diff
@@ -345,12 +345,12 @@ func (s *Service4Value) GetBackendID() loadbalancer.BackendID {
 
-func (s *Service4Value) GetLbAlg() uint8 {
-	return uint8(uint32(s.BackendID) >> 24)
+func (s *Service4Value) GetLbAlg() loadbalancer.SVCLoadBalancingAlgorithm {
```
please verify that those functions are properly tested :)

DamianSawicki replied:
Haha, I was already working on it :D

DamianSawicki replied:
Done, PTAL.

kl52752 reviewed (commented):
Thank you for your work! One nit otherwise looks good to me! 
/lgtm

DamianSawicki committed:
fix: SetLbAlg and SetSessionAffinityTimeoutSec

This corrects the implementation of SetLbAlg() and
SetSessionAffinityTimeoutSec(), adds GetSessionAffinityTimeoutSec()
for completeness, and also adds a test for all these.

Fixes: 1b2bed77dad58e1c98d801762b79262a40a3bf2c
Signed-off-by: Damian Sawicki <dsawicki@google.com>

DamianSawicki commented on pkg/loadbalancer/experimental/bpf_reconciler.go:
```diff
@@ -18,9 +18,11 @@ import (
 	"golang.org/x/sys/unix"
```
> > Conformance Kind Envoy Embedded / Installation and Connectivity Test (pull_request) Failing after 16m
> 
> ```
> 📋 Test Report [cilium-test-5]
> ❌ 1/13 tests failed (1/196 actions), 6 tests skipped, 0 scenarios skipped:
> Test [client-egress-l7-set-header-port-range]:
>   ❌ client-egress-l7-set-header-port-range/pod-to-pod-with-endpoints/curl-ipv4-2-auth-header-required: cilium-test-5/client3-795488bf5-8lcsb (10.244.0.214) -> curl-ipv4-2-auth-header-required (10.244.1.250:8080)
> ```
> 
> I see very similar test failures for an empty PR https://github.com/cilium/cilium/actions/runs/12429666814/job/34703608166?pr=36744, so it's most likely some flakiness or regression on `main`.

Yeah, it succeeded now (which is probably not related to changes introduced in the meantime), although 5 other sub-tests failed

Status | Package | Time (seconds)
-- | -- | --
❌ | client-egress-l7-tls-headers-sni | 194.916
❌ | client-egress-tls-sni | 441.098
❌ | no-policies | 635.625
❌ | seq-client-egress-l7-tls-headers | 195.541
❌ | seq-client-egress-l7-tls-headers-port-range | 195.290

(which I suppose succeeded before).

brb commented on pkg/maps/lbmap/ipv4.go:
```diff
@@ -328,8 +328,13 @@ func (s *Service4Value) GetFlags() uint16 {
 	// lb4_service.affinity_timeout field. Also, for the master entry the
-	// LB algorithm was set beforehand, so we need to binary OR here.
-	s.BackendID |= t
+	// LB algorithm can be set independently, so we need to preseve the
```
I was not aware of https://github.com/cilium/cilium/pull/36204/, but I think it might make sense to change the method to return an error if a given affinity timeout `t` is > `0x00ffffff`. Then the error should be logged with level=warn. Otherwise, we might silently configure a different than expected by a user timeout.

DamianSawicki replied:
> I think it might make sense to change the method to return an error if a given affinity timeout `t` is > `0x00ffffff`.

Done. 

> Then the error should be logged with level=warn.

Errors were already propagated in `pkg/loadbalancer/experimental` and finally logged at `warn`, but error handling was more complicated in the non-experimental case (with call stack reaching for instance `api/v1/server/restapi/service/put_service_id.go` and `pkg/service/health_check.go` without logging), so as you suggested I added logging in `pkg/maps/lbmap/lbmap.go`.

tommyp1ckles commented on pkg/loadbalancer/experimental/testdata/nodeport-explicit-maglev.txtar:
```diff
@@ -0,0 +1,285 @@
+#! --enable-experimental-lb --bpf-lb-algorithm-annotation
```
suggestion: could we put these test data changes into a seperate commit to make the changes easier to understand?

DamianSawicki replied:
Done. I increased the number of commits from 2 to 4:

- One extra commit is for the change of the format in `DumpLBMaps()` (this commit goes before the main feature commit), and
- the other is for new tests for the feature (the file `pkg/loadbalancer/experimental/testdata/nodeport-explicit-maglev.txtar` that this comment is on, the twin file `pkg/loadbalancer/experimental/testdata/nodeport-explicit-random.txtar`, and the new test cases `perServiceAlgorithmCases` in `pkg/loadbalancer/experimental/bpf_reconciler_test.go`).

The remaining test changes in the main feature commit are necessary to keep the invariant that all tests pass for every commit.

Hope that helps!

tommyp1ckles commented on pkg/maps/lbmap/ipv4v6_test.go:
```diff
@@ -13,10 +14,11 @@ import (
+		alg2        loadbalancer.SVCLoadBalancingAlgorithm = 2
+		time1       uint32                                 = 8
+		time2       uint32                                 = 16
+		invalidTime uint32                                 = uint32(math.Exp2(24))
```
nit:  we could just do 1 << 24 instead of casting out of the float exp value.

DamianSawicki replied:
Good idea! I was unsuccessfully looking up an exponentiation operator in Golang, but binary shift also works great here 😊  Done.

tommyp1ckles commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -631,7 +631,10 @@ func updateMasterService(fe ServiceKey, v ServiceValue, activeBackends, quaranti
 	if sessionAffinity {
-		v.SetSessionAffinityTimeoutSec(sessionAffinityTimeoutSec)
+		if err := v.SetSessionAffinityTimeoutSec(sessionAffinityTimeoutSec); err != nil {
+			log.Warn("Failure in updateMasterService due to error from SetSessionAffinityTimeoutSec: %w")
```
unused/invalid format specifier in this log: `%w`

DamianSawicki replied:
Thanks, fixed!

DamianSawicki committed:
experimental: Add LBALG to DumpLBMaps

This adds LBALG to DumpLBMaps in preparation for adding support for
Per-service Load Balancing Algorithm Selection (CFP-34577) to
StateDB-based load balancing.

Signed-off-by: Damian Sawicki <dsawicki@google.com>

DamianSawicki committed:
experimental: Per service lb algorithm selection

This adds the feature from CFP-34577 to StateDB-based load balancing.

Signed-off-by: Damian Sawicki <dsawicki@google.com>

DamianSawicki committed:
experimental, test: Per svc lb algorithm selection

This commit adds new tests for the support of Per-service Load Balancing
Algorithm Selection (CFP-34577) in StateDB-based load balancing
(isolated to a separate commit as suggested in code review).

Signed-off-by: Damian Sawicki <dsawicki@google.com>

tommyp1ckles commented on pkg/maps/lbmap/ipv4.go:
```diff
@@ -325,11 +325,20 @@ func (s *Service4Value) GetFlags() uint16 {
+	// LB algorithm can be set independently, so we need to preseve the
+	// first 8 bits and only assign to the latter 24 bits.
+	if t > 0x00ff_ffff {
+		return fmt.Errorf("session affinity timeout %d does not fit into 24 bits (is larger than 16777215)", t)
```
we could probably just define `0x00ff_ffff` as a constant given its widespread use.

DamianSawicki replied:
Good idea, I'll define `0x00ff_ffff` and `0xff00_0000` as constants.

DamianSawicki replied:
Done.

tommyp1ckles commented on pkg/loadbalancer/experimental/config.go:
```diff
@@ -40,28 +40,33 @@ type TestConfig struct {
 	// flags move away from option.DaemonConfig and can thus be set directly.
 	NodePortAlg string `mapstructure:"node-port-algorithm"`
+
+	LoadBalancerAlgorithmAnnotation bool `mapstructure:"bpf-lb-algorithm-annotation"`
```
Usually, flags that toggle on some functionality are named "enable-...", that might make the purpose of this flag more clear?

DamianSawicki replied:
Thanks for reviewing! The flag name is identical as in https://github.com/cilium/cilium/blob/62e26d498de76c734e8380ad9c1df7caf12a9d1a/pkg/option/config.go#L269, so I think I shouldn't modify it in this PR, but I can add a clarifying comment (similar to the comment above `NodePortAlg`).

DamianSawicki replied:
Comment added.

tommyp1ckles reviewed (approved):
@cilium/sig-k8s changes lgtm, just a few more minor points/suggestions.

brb reviewed (approved):
Thanks!

sayboras reviewed (approved):
Looks good for files under ciliumenvoyconfig

joamaki commented on pkg/loadbalancer/experimental/testdata/dualstack-maglev.txtar:
```diff
@@ -283,39 +283,39 @@ REV: ID=6 ADDR=1.1.1.1:30181
 REV: ID=8 ADDR=1.1.1.1:32757
 REV: ID=9 ADDR=<zero>
-SVC: ID=1 ADDR=10.96.207.99:80 SLOT=0 BEID=0 COUNT=2 QCOUNT=0 FLAGS=ClusterIP+non-routable
+SVC: ID=1 ADDR=10.96.207.99:80 SLOT=0 BEID=33554432 COUNT=2 QCOUNT=0 LBALG=maglev FLAGS=ClusterIP+non-routable
```
Why are all the backend ids messed up?

DamianSawicki replied:
One field is reused for all: backend id, session affinity timeout, and lb algorithm https://github.com/cilium/cilium/blob/d6d40811c35ee7f8f79fbaa9e72af22fe460679f/bpf/lib/common.h#L1085-L1097 (and also for l7_lb_proxy_port, but it seems irrelevant here). When session affinity is used, it is already visible in BEID https://github.com/cilium/cilium/blob/d6d40811c35ee7f8f79fbaa9e72af22fe460679f/pkg/loadbalancer/experimental/bpf_reconciler_test.go#L678, so I didn't attempt to change this behaviour. Large numbers in diffs come from the fact that a binary shift is used to store both session affinity timeout and lb algorithm in one variable. In `bpf_reconciler_test.go` there is a comment on it: https://github.com/cilium/cilium/blob/edd4db314d73b4c5f7c0fe81c6b8072a396599c4/pkg/loadbalancer/experimental/bpf_reconciler_test.go#L1065.

joamaki replied:
Let's fix that by extracting out the actual backend id so one can match SVC lines with BE lines, and let's show the other info separately. For the master entry we could just omit `BEID=` and instead show e.g. `AFFTIMEOUT=...` etc.

Or I suppose alternatively we could show it raw but in hex so it's easier to read.

DamianSawicki replied:
I went for the first suggestion: I'm using either `BEID=` or other names like `LBALG=` depending on the actual meaning of the union field. PTAL.

DamianSawicki replied:
Resolving as Jussi already approved.

joamaki reviewed (changes requested):
Blocking until the `BEID` fixed in the lbmap dumps.

DamianSawicki committed:
experimental: BEID stringified based on union case

This changes format of BPF dumps in tests. BackendID is stored in a
union type together with affinity timeout, loadbalancing algorithm, and
L7 proxy port, and now the string depends on the union case.

Signed-off-by: Damian Sawicki <dsawicki@google.com>

joamaki commented on pkg/ciliumenvoyconfig/testdata/anyport.txtar:
```diff
@@ -150,9 +150,9 @@ BE: ID=1 ADDR=10.244.1.1:8080 STATE=active
+SVC: ID=1 ADDR=10.96.50.104:80 SLOT=0 L7Proxy=59395 COUNT=1 QCOUNT=0 FLAGS=ClusterIP+non-routable+l7-load-balancer
 SVC: ID=1 ADDR=10.96.50.104:80 SLOT=1 BEID=1 COUNT=0 QCOUNT=0 FLAGS=ClusterIP+non-routable+l7-load-balancer
-SVC: ID=2 ADDR=10.96.50.104:25 SLOT=0 BEID=59395 COUNT=1 QCOUNT=0 FLAGS=ClusterIP+non-routable+l7-load-balancer
+SVC: ID=2 ADDR=10.96.50.104:25 SLOT=0 L7Proxy=59395 COUNT=1 QCOUNT=0 FLAGS=ClusterIP+non-routable+l7-load-balancer
```
We should convert the endianess and show the correct port number, e.g. this should read `L7Proxy=1000`.

DamianSawicki replied:
This seems to be fixed now.

joamaki commented on pkg/loadbalancer/experimental/service.go:
```diff
@@ -130,6 +131,7 @@ func (svc *Service) TableHeader() []string {
 		"HealthCheckNodePort",
 		"LoopbackHostPort",
 		"SourceRanges",
+		"ExplicitLBAlgorithm",
```
Hmm, starting to wonder whether we should start collapsing the more niche things into a single column soon. I think this is fine for now, but I'm guessing at some point the `db/show services` starts to look a bit overwhelming.

DamianSawicki replied:
It's getting lengthy, indeed.

joamaki commented on pkg/loadbalancer/experimental/service.go:
```diff
@@ -166,6 +168,7 @@ func (svc *Service) TableRow() []string {
 		strconv.FormatUint(uint64(svc.HealthCheckNodePort), 10),
 		showBool(svc.LoopbackHostPort),
 		showSourceRanges(svc.SourceRanges),
+		svc.Annotations[annotation.ServiceLoadBalancingAlgorithm],
```
For accessing such a core annotation like this I would add a getter method to `Service`

DamianSawicki replied:
Added.

joamaki commented on pkg/loadbalancer/experimental/test_utils.go:
```diff
@@ -145,6 +145,20 @@ func DumpLBMaps(lbmaps LBMaps, feAddr loadbalancer.L3n4Addr, sanitizeIDs bool, c
+			return fmt.Sprintf("BEID=%s", sanitizeID(svcValue.GetBackendID(), sanitizeIDs))
+		}
+		if loadbalancer.ServiceFlags(svcValue.GetFlags()).IsL7LB() {
+			return fmt.Sprintf("L7Proxy=%d", svcValue.GetBackendID())
```
Here we do have `svcValue` in host endianess, so not quite sure why `GetBackendID()` here returns the L7 proxy port in the wrong endianess. Could you dig into that a bit and see what's up and convert it correctly here?

DamianSawicki replied:
I added a getter for this which does the same thing as the setter, so hopefully it's correct, PTAL.

> so not quite sure why GetBackendID() here returns the L7 proxy port in the wrong endianess

Yourself from August had such a comment on this topic:
https://github.com/cilium/cilium/blob/13cc1f1944351efe256862ac68bde4eada426973/pkg/loadbalancer/experimental/bpf_reconciler_test.go#L427-L428

DamianSawicki replied:
Resolving as Jussi already approved.

DamianSawicki committed:
lbmap: Add GetL7LBProxyPort to ServiceValue

This is not to avoid using GetBackendID for L7 Proxy in order to
handle endianness in the same way as in the setter.

Signed-off-by: Damian Sawicki <dsawicki@google.com>

DamianSawicki committed:
experimental: Add svc lb alg annotation getter

As requested in code review, this adds method
GetLBAlgorithmAnnotation() string
to experimental.Service that explicitly provides the value of the
relevant annotation.

This is in contrast with the functions
getAnnotationServiceLoadBalancingAlgorithm()
GetAnnotationServiceLoadBalancingAlgorithm()
in pkg/k8s/service.go that take into account default algorithms and
return erors.

Using this occassion to modify the signature of
GetAnnotationServiceLoadBalancingAlgorithm()
in pkg/k8s/service.go to pass the default algorithm with its
specialized type instead of the string type.

Signed-off-by: Damian Sawicki <dsawicki@google.com>
